### PR TITLE
First Step Towards Movie List

### PR DESCRIPTION
**Pull request by augustdolan**
I'm opening this commit to make comments on @devontriley's work

### CODE REVIEW

**Comment by augustdolan** [comment]
Just a comment on the ui: 

I wish I could whip up something this nice looking so quick - super nice. 

I might change the default value to "all" for genre select, and add an "all" for both.

File: src/MovieSearch/DisplayResults.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    return (
+        <div className="results mt-3">
+            {!searchResults.length || typeof searchResults == 'string'
+                ?
+                    <p>No results</p>
+                :
+                    <ul>
+                        {searchResults.map(result => {
+                            return (
+                                <li key={result.id}>{result.title}</li>
```
Comment (augustdolan) [inline src/MovieSearch/DisplayResults.js:14]: Often times folks will actually split this up into two components. So for instance:

<MovieList>, which for each result in searchResults would return a <MovieListEntry>

Generally, I would try to make your components as semantically sound as possible. "Results" doesn't tell me anything about my component, but <MovieList> tells me its a list of movies

File: src/MovieSearch/GenreSelect.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+const getGenres = () => {
+    return axios.get(`https://api.themoviedb.org/3/genre/movie/list?api_key=b3a999b7703140535e95baeff2e338fa&language=en-US`)
+        .then(res => {
+            return res.data.genres;
+        });
+}
+
+const GenreSelect = () =>
+{
+    const [genres, setGenres] = useState([]);
```
Comment (augustdolan) [inline src/MovieSearch/GenreSelect.js:13]: Although we need to populate genres _after_ startup, it isn't _stateful_ in the sense that we expect it to change. Genres are static - once we got'em, we got'em. I'd get rid of the state, just save this to a const genres and call it.

File: src/MovieSearch/GenreSelect.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        getGenres()
+            .then(data => {
+                setGenres(data);
+            });
+    }, []);
+
+    return (
+        <div className="field-group mb-2">
+            <select className="form-control">
+                {genres.map(genre =>
```
Comment (augustdolan) [inline src/MovieSearch/GenreSelect.js:26]: same thing here. We might have `<Genres>`, which maps info to a `<GenreEntry>`

File: src/MovieSearch/MovieSearch.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import React, { useState } from 'react';
+import MovieSearchForm from "./MovieSearchForm";
+import DisplayResults from "./DisplayResults";
+
+const MovieSearch = () =>
+{
+    const [searchResults, setSearchResults] = useState('No results');
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearch.js:7]: Instead of initializing at 'No results' and then checking if there is an empty array || a string, you can just initialize this at `useState([])` and have your check be: 

`searchResults ? thing I do when I have results : <p>No results!</p>`

Comment (augustdolan) [reply src/MovieSearch/MovieSearch.js:7]: This is really helpful because now when people read your initial state, they know what kind of data to expect. When I saw 'No Results' I think to myself "ok, this state is for string values"

File: src/MovieSearch/MovieSearchBar.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+const MovieSearchBar = ({onTextChange}) =>
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearchBar.js:3]: Is this an [Uncontrolled Component?](https://reactjs.org/docs/uncontrolled-components.html#the-file-input-tag)

If so I would _highly_ recommend against that. It is best to put our source of truth in one place - which in this case is our React framework and the vDOM. Get some state in here to [control the component's](https://reactjs.org/docs/forms.html#controlled-components) form information. It is generally safer, and much more clear what's going on

File: src/MovieSearch/MovieSearchForm.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import React from 'react';
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearchForm.js:1]: It got really hard to follow your component tree as soon as I realized MovieSearch was the name of a file AND a folder, and there were a ton of similar sounding names: 

MovieSearchForm
MovieSearchBar
MovieSearch

I'm left kinda scratching my head and digging around for things. I think there's a better way to compartmentalize this. I'd be interested to see you draw out the parent => child tree diagram of this repo

File: src/MovieSearch/MovieSearchForm.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+const searchMovies = (keyword) =>
+{
+    return axios.get(`https://api.themoviedb.org/3/search/movie?api_key=b3a999b7703140535e95baeff2e338fa&query=${keyword}`)
+        .then(res => {
+            return res.data.results;
+        });
+}
+
+const filterResultsByGenre = (results, genre) =>
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearchForm.js:14]: Does the API not let us query the resource collection by keyword && genre? It would be better for the client if we can let the server handle parsing out the genre we want

File: src/MovieSearch/MovieSearchForm.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+{
+    let genreInt = parseInt(genre);
+
+    let filteredResults = results.filter(function(item)
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearchForm.js:18]: I might update semantics here. 

`filterMoviesByGenre = (movies, genre)`

`movies.filter(function(movie)`

Now everything is very clear. With `results` and `items`, I'm not sure what those represent - they could be anything.

File: src/MovieSearch/MovieSearchForm.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+{
+    return axios.get(`https://api.themoviedb.org/3/search/movie?api_key=b3a999b7703140535e95baeff2e338fa&query=${keyword}`)
+        .then(res => {
+            return res.data.results;
+        });
+}
+
+const filterResultsByGenre = (results, genre) =>
+{
+    let genreInt = parseInt(genre);
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearchForm.js:16]: Are we type checking if we somehow get a non enumerated option? That is if `'action'` sneaks in, we want to say its NaN? 

It looks like we are checking the index of whatever is passed in, so if we get a nonnumber it would fail either way, right?

File: src/MovieSearch/MovieSearchForm.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    });
+
+    return filteredResults;
+}
+
+const MovieSearchForm = ({setSearchResults}) =>
+{
+    const handleSubmission = (e) =>
+    {
+        const genreSelect = e.target[0];
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearchForm.js:30]: Ok, I think this lines up with my assumption that you are using uncontrolled components. 

The thing is that with React, _we're_ controlling the DOM. It is best to keep it that way unless we have a good reason not to.

File: src/MovieSearch/MovieSearchForm.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        const textInput = e.target[1];
+
+        searchMovies(textInput.value)
+            .then(data =>
```
Comment (augustdolan) [inline src/MovieSearch/MovieSearchForm.js:34]: What kind of data is it? I'd change the name for that. `movies` and `let filteredMovies`

File: src/Watchlist/Watchlist.js [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import React, { useState } from 'react';
+import GenreSelect from "../MovieSearch/GenreSelect";
+import DisplayResults from "../MovieSearch/DisplayResults";
+
+const Watchlist = () => {
```
Comment (augustdolan) [inline src/Watchlist/Watchlist.js:5]: This isn't useful yet, right? No need to review this since we can't really see any results moved there yet?

File: src/MovieSearch/DisplayResults.js (hunk 1 above)
Comment (devontriley) [reply src/MovieSearch/DisplayResults.js:14]: This makes alot of sense and is perfect for what I want to make next: a way to add results to your watchlist - which clearly wants components for each child at that point.

File: src/MovieSearch/GenreSelect.js (hunk 2 above)
Comment (devontriley) [reply src/MovieSearch/GenreSelect.js:13]: Okay - I was also thinking this too but I couldn't figure out how to do it correctly. I can't use *useEffect* if I'm not using state, so how can I load the async fetching of the genres after the initial load of the component?

File: src/MovieSearch/DisplayResults.js (hunk 1 above)
Comment (augustdolan) [reply src/MovieSearch/DisplayResults.js:14]: Yes! And that's why we do stuff like this! Great point and great idea :)

File: src/MovieSearch/GenreSelect.js (hunk 2 above)
Comment (augustdolan) [reply src/MovieSearch/GenreSelect.js:13]: You actually can use useEffect without state! useEffect is only about the _lifecycle_ of your component. You're only describing things you want to have happen during certain parts of that lifecycle. Any JavaScript is valid in there!

File: src/MovieSearch/MovieSearchForm.js (hunk 6 above)
Comment (devontriley) [reply src/MovieSearch/MovieSearchForm.js:1]: Screen Shot 2021-04-14 at 5 52 33 PM

Comment (devontriley) [reply src/MovieSearch/MovieSearchForm.js:1]: I definitely agree about the naming though - I'll clean it up. For something like `MovieSearchBar`, can that just be simplified to `SearchBar` or is keeping it more specific better?

File: src/MovieSearch/MovieSearchForm.js (hunk 7 above)
Comment (devontriley) [reply src/MovieSearch/MovieSearchForm.js:14]: It does not :( I thought this was a pretty good solution though

File: src/MovieSearch/MovieSearchForm.js (hunk 9 above)
Comment (devontriley) [reply src/MovieSearch/MovieSearchForm.js:16]: Not sure about this. I know for a fact the api only uses integers for the Genre ID. Are you saying the function should check for non integers anyway and handle accordingly?

File: src/MovieSearch/MovieSearchForm.js (hunk 7 above)
Comment (augustdolan) [reply src/MovieSearch/MovieSearchForm.js:14]: Are we sure there isn't an endpoint? If so, yeah I agree this is a great solution

Comment (augustdolan) [reply src/MovieSearch/MovieSearchForm.js:14]: You could even go buck wild, set up a server, have the server call the api, process the results, and then send to the client 👀

File: src/MovieSearch/MovieSearchForm.js (hunk 9 above)
Comment (augustdolan) [reply src/MovieSearch/MovieSearchForm.js:16]: Just trying to figure out why you parseInt genre

File: src/MovieSearch/GenreSelect.js (hunk 2 above)
Comment (devontriley) [reply src/MovieSearch/GenreSelect.js]: Ohhhh okay, I didn't realize that!
